deduplication-document-spark/deduplication-document-spark-impl/src/main/scala/pl/edu/icm/coansys/document/deduplication/DeduplicateDocuments.scala

```diff
@@ -0,0 +1,420 @@
+        println("No config, aborting.")
+        return
+    }
+
+    println("Creating context...")
+
+    //required to operate protobuf correctly
+    val conf = new SparkConf()
+      .setAppName("Document deduplication")
+      .set("spark.serializer", "org.apache.spark.serializer.KryoSerializer")
+      .set("spark.kryo.registrator", "pl.edu.icm.coansys.document.deduplication.DocumentWrapperKryoRegistrator")
+
+    val sc = new SparkContext(conf)
+
+    println("Created context...")
+    //    sc.getConf.getAll.foreach(x => println(x._1 + ": " + x._2))
+    val inputDocuments = cfg.inputFile
+    // for pasting into console:
+    //  val inputDocuments = "/user/kura/curr-res-navigator/hbase-sf-out/DOCUMENT"
+    //  val inputDocuments = "/user/kura/curr-res-navigator-no-blogs/hbase-sf-out/DOCUMENT"
+    val outputDocuments = cfg.outputFile
+    
+        
+    //load documents 
+    val wrappers = loadDocuments(sc, inputDocuments, cfg.filterInvalidDocuments, cfg.removeDuplicateDocuments)
+    val initialSize = wrappers.count
+    println(f"Starting processing with $initialSize documents.")
+
+    val initialGroups = prepareInitialClustering(wrappers, cfg.keySizeMin, cfg.keySizeMax, cfg.clusterSizeMax)
+    initialGroups.persist
+
+    val clustersToDeduplicate = initialGroups.filter(t => t._2.size > 1)
+    val initialClusterCount = clustersToDeduplicate.count
+    //TODO: some statistics here on cluster, would be useful.
+
+    val tiledTasks = clustersToDeduplicate.flatMap(p => CartesianTaskSplit.parallelizeCluster(p._1, p._2, cfg.tileSize))
+    
+    tiledTasks.persist
+    val tileCount = tiledTasks.count;
+
+    println(f"Prepared $initialClusterCount clusters, and then split it to $tileCount tiles")
+    
+    //build (clusterId, Seq(docId)) rdd:    
+    val partialEqualityClusters = tiledTasks.flatMap(
+      task => {
+        val t0 = java.lang.System.currentTimeMillis()
+        val comparator = buildWorkComparator
+        val res = task.processPairs((a: DocumentWrapper, b: DocumentWrapper) =>
+          comparator.isDuplicate(a.getDocumentMetadata, b.getDocumentMetadata, null))
+        val time = (java.lang.System.currentTimeMillis() - t0) / 1000.0
+        //useful for identification of possible problem.
+        log.info(f"Finishing tile task ${task.taskId} in $time%.4f sec")
+        res.map((task.clusterId, _))
+      }
+    )
+    
+    val finalClusters = partialEqualityClusters.mapValues(List(_)).
+      reduceByKey(_ ++ _). //one long list of lists of ids for each cluster
+      map(pair => {
+        val t0 = java.lang.System.currentTimeMillis()
+        val res = CartesianTaskSplit.coalesceClusters(pair._2.asJava)
+        val tt = System.currentTimeMillis() - t0;
+        val clusterSize = pair._2.size
+        log.info(f"Finished tile coalesce task. (Cluster,time[s], size): ${pair._1}, ${tt / 1000.0}, ${pair._2.size}")
+        (pair._1, res)
+      }).
+      flatMap( //build proper ids for equality clusters.
```

Comment:
Pattern matching should be used in map & flatMap just like described previously.
Additionally longer names than `cid` & `cl` (`clusterId` & `cluster`?) would be clearer.